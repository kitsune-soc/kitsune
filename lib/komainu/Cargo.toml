[package]
name = "komainu"
authors.workspace = true
edition.workspace = true
version.workspace = true
license = "MIT OR Apache-2.0"

[[bench]]
name = "basic_auth"
harness = false

[[bench]]
name = "pkce"
harness = false

[dependencies]
base64-simd.workspace = true
bytes.workspace = true
compact_str.workspace = true
http.workspace = true
http-body.workspace = true
http-body-util.workspace = true
indexmap.workspace = true
itertools.workspace = true
memchr.workspace = true
serde.workspace = true
serde_urlencoded.workspace = true
sha2.workspace = true
simdutf8.workspace = true
sonic-rs.workspace = true
strum.workspace = true
subtle.workspace = true
thiserror.workspace = true
tracing.workspace = true
url.workspace = true

[dev-dependencies]
divan.workspace = true
futures-test.workspace = true
headers.workspace = true
insta.workspace = true
rstest.workspace = true
serde_test.workspace = true

[lints]
workspace = true
