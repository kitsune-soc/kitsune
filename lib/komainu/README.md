# komainu

![](https://www.japan-experience.com/sites/default/files/images/content_images/komainu20191.jpg)

Standards compliant generic OAuth2 server library for async Rust webservices.

Relevant standards:

- [RFC 6749](https://datatracker.ietf.org/doc/html/rfc6749)
- [RFC 7636](https://datatracker.ietf.org/doc/html/rfc7636)

## Goals

Make it easy for anyone to implement a full OAuth2 provider for their webservice.

## License

This project is licensed under either the MIT or Apache-2.0 license, at your choosing.
