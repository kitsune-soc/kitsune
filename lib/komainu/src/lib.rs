#[macro_use]
extern crate tracing;

use self::{error::Error, flow::pkce, scope::Scope};

pub use self::primitive::{Authorization, Client, Request};

mod primitive;

pub mod code_grant;
pub mod error;
pub mod extract;
pub mod flow;
pub mod params;
pub mod scope;

pub struct AuthInstruction<'a, 'b> {
    pub client: &'b Client<'a>,
    pub pkce_payload: Option<&'b pkce::Payload<'a>>,
    pub scopes: &'b Scope,
}

pub trait ClientExtractor {
    fn extract(
        &self,
        client_id: &str,
        client_secret: Option<&str>,
    ) -> impl Future<Output = Result<Client<'_>, Error>> + Send;
}
