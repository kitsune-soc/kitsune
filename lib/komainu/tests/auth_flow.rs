use self::fixtures::{generate_secret, Fixture, RNG_SEED};
use komainu::{flow, scope::Scope, ClientExtractor};

mod fixtures;

#[futures_test::test]
async fn success_basic() {
    fastrand::seed(RNG_SEED);

    let fixtures = Fixture::generate();
    let code = generate_secret();
    let client = fixtures
        .client_extractor
        .extract("client_1", None)
        .await
        .unwrap();

    fixtures.auth_storage.insert(komainu::Authorization {
        code: code.clone().into(),
        client: client.clone(),
        pkce_payload: None,
        scopes: Scope::new(),
        user_id: "user_id".into(),
    });

    let body = sonic_rs::json!({
        "grant_type": "authorization_code",
        "code": code,

        "client_id": client.client_id,
        "client_secret": client.client_secret,
        "redirect_uri": client.redirect_uri,
    });
    let body = sonic_rs::to_string(&body).unwrap();

    let req = http::Request::builder()
        .header(http::header::CONTENT_TYPE, "application/json")
        .body(body)
        .unwrap();
    let req = komainu::Request::read_from(req).await.unwrap();

    let response =
        flow::authorization::perform(&req, &fixtures.client_extractor, &fixtures.auth_flow)
            .await
            .unwrap();

    insta::assert_json_snapshot!(response);
}
