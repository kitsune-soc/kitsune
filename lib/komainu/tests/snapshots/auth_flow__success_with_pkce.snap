---
source: lib/komainu/tests/auth_flow.rs
expression: response
---
{
  "access_token": "zisudtbgojtcgfsh",
  "token_type": "Bearer",
  "refresh_token": "mcnfelbaydmjbddc",
  "expires_in": 1337
}
